llama_index/vector_stores/singlestoredb.py

```diff
@@ -0,0 +1,246 @@
+            metadata_field: str = "metadata",
+            vector_field: str = "vector",
+            pool_size: int = 5,
+            max_overflow: int = 10,
+            timeout: float = 30,
+            **kwargs: Any,
+        ) -> None:
+        """Init params."""
+        self.table_name = table_name
+        self.content_field = content_field
+        self.metadata_field = metadata_field
+        self.vector_field = vector_field
+        self.pool_size = pool_size
+        self.max_overflow = max_overflow
+        self.timeout = timeout
+
+        self.connection_kwargs = kwargs
+        self.connection_pool = QueuePool(
+            self._get_connection,
+            pool_size=self.pool_size,
+            max_overflow=self.max_overflow,
+            timeout=self.timeout,
+        )
+        
+        self._create_table()
+
+
+    @property
+    def client(self) -> Any:
+        """Return SingleStoreDB client."""
+        return self._get_connection()
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "SingleStoreVectorStore"
+
+    def _get_connection(self) -> Any:
+        try:
+            import singlestoredb as s2
+        except ImportError:
+            raise ImportError(
+                "Could not import singlestoredb python package. "
+                "Please install it with `pip install singlestoredb`."
+            )
+        return s2.connect(**self.connection_kwargs)
+    
+    def _create_table(self) -> None:
+        conn = self.connection_pool.connect()
+        cursor = conn.cursor()
+        try:
+            cur = conn.cursor()
+            try:
+                cur.execute(
+                    f"""CREATE TABLE IF NOT EXISTS {self.table_name}
+                    ({self.content_field} TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci,
+                    {self.vector_field} BLOB, {self.metadata_field} JSON);"""
+                )
+            finally:
+                cur.close()
+        finally:
+            conn.close()
+
+    def add(self, nodes: List[BaseNode]) -> List[str]:
+        """Add nodes to index.
+
+        Args
+            nodes: List[BaseNode]: list of nodes with embeddings
+
+        """
+        conn = self.connection_pool.connect()
+        cursor = conn.cursor()
+        try:
+            for node in nodes:
+                embedding = node.get_embedding()
+                metadata = node_to_metadata_dict(
+                    node, remove_text=True, flat_metadata=self.flat_metadata
+                )
+                cursor.execute(
+                        "INSERT INTO {} VALUES (%s, JSON_ARRAY_PACK(%s), %s)".format(
+                            self.table_name
+                        ),
+                        (
+                            node.get_content(metadata_mode=MetadataMode.NONE) or "",
+                            "[{}]".format(",".join(map(str, embedding))),
+                            json.dumps(metadata),
+                        ),
+                    )
```

Comment:
nit: is there performance concern here inserting nodes in a loop instead of in bulk?

Reply:
Nope, should be a neglible difference